op-service/sources/l1_beacon_client.go

```diff
@@ -0,0 +1,157 @@
+package sources
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/crypto/kzg4844"
+
+	"github.com/ethereum-optimism/optimism/op-service/client"
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+)
+
+type L1BeaconClient struct {
+	cl client.HTTP
+
+	initLock     sync.Mutex
+	timeToSlotFn TimeToSlotFn
+}
+
+// NewL1BeaconClient returns a client for making requests to an L1 consensus layer node.
+func NewL1BeaconClient(cl client.HTTP) *L1BeaconClient {
+	return &L1BeaconClient{cl: cl}
+}
+
+func (cl *L1BeaconClient) apiReq(ctx context.Context, dest any, method string) error {
+	headers := http.Header{}
+	headers.Add("Accept", "application/json")
+	resp, err := cl.cl.Get(ctx, method, headers)
+	if err != nil {
+		return fmt.Errorf("%w: http Get failed", err)
+	}
+	if resp.StatusCode != http.StatusOK {
+		errMsg, _ := io.ReadAll(resp.Body)
+		_ = resp.Body.Close()
+		return fmt.Errorf("failed request with status %d: %s", resp.StatusCode, string(errMsg))
+	}
+	if err := json.NewDecoder(resp.Body).Decode(dest); err != nil {
+		_ = resp.Body.Close()
+		return err
+	}
+	if err := resp.Body.Close(); err != nil {
+		return fmt.Errorf("%w: failed to close response body", err)
+	}
+	return nil
+}
+
+type TimeToSlotFn func(timestamp uint64) uint64
+
+// GetTimeToSlotFn returns a function that converts a timestamp to a slot number.
+func (cl *L1BeaconClient) GetTimeToSlotFn(ctx context.Context) (TimeToSlotFn, error) {
+	cl.initLock.Lock()
+	defer cl.initLock.Unlock()
+	if cl.timeToSlotFn != nil {
+		return cl.timeToSlotFn, nil
+	}
+
+	var genesisResp eth.APIGenesisResponse
+	if err := cl.apiReq(ctx, &genesisResp, "eth/v1/beacon/genesis"); err != nil {
+		return nil, err
+	}
+
+	var configResp eth.APIConfigResponse
+	if err := cl.apiReq(ctx, &configResp, "eth/v1/config/spec"); err != nil {
+		return nil, err
+	}
+
+	cl.timeToSlotFn = func(timestamp uint64) uint64 {
```

Comment:
Should we check for underflow here and have the function return an error as well for that case?

Reply:
makes sense, done